### PR TITLE
Added device event subscription

### PR DESCRIPTION
**Pull request by markoImake**
Hi Eddy
I was working on my own nativescript-particle plugin until I saw you recently published this one. I added the additional features that I was using into your plugin. These include:
1. loginWithToken - this is needed for managing shadow accounts when implementing Two-Legged Authentication (https://docs.particle.io/guide/how-to-build-a-product/authentication/#two-legged-authentication)
2. isAuthenticated and accessToken methods to make it easier to manage shadow accounts
3. subscribe to (and unsubscribe from) events on a TNSParticleDevice
4. Particle Device Setup Wizard for ios, using particle 'ParticleSetup' sdk (https://github.com/particle-iot/particle-setup-ios)

### CODE REVIEW

File: src/particle.android.ts [hunk 1]
```diff
@@ -31,17 +33,36 @@ export class Particle implements TNSParticleAPI {
     });
   }
 
+  public loginWithToken(token: string): void {
+    ParticleCloudSDK.getCloud().setAccessToken(token);
+    if (global["TNS_WEBPACK"]) {
+      const WorkerScript = require("nativescript-worker-loader!./particle-worker.js");
+      worker = new WorkerScript();
+      const EventWorkerScript = require("nativescript-worker-loader!./particle-event-worker.js");
```
Comment (EddyVerbruggen) [inline src/particle.android.ts]: Do we need to do this for the other login method as well? Or better still: can we start the eventworker lazily when a subscribe call is made?

File: src/particle.android.ts [hunk 2]
```diff
@@ -87,4 +113,55 @@ export class Particle implements TNSParticleAPI {
...
+  public isAuthenticated(): Boolean {
+    return ParticleCloudSDK.getCloud().isLoggedIn();
+}
+  
+  public accessToken(): string {
+    return ParticleCloudSDK.getCloud().getAccessToken();
+  }
+
+  public startDeviceSetupWizard(cb:any): void {
+    // stub for startDeviceSetupWizard
```
Comment (EddyVerbruggen) [inline src/particle.android.ts:169]: For my info: there's no device setup wizard for Android?

**Review by EddyVerbruggen — CHANGES_REQUESTED** [review]
Hi! This is honestly excellent work! I'm very glad you decided to merge your code with the repo. And I'd be happy yo add you as a collaborator to the repo if you want to.

I've asked 2 questions, please check them out when you have time.

I'd be happy to test the updated demo app and add docs for these new features.

Thanks,
Eddy

File: src/particle.android.ts (hunk 2 above)
Comment (markoImake) [reply src/particle.android.ts:169]: There is a device setup wizard available for Android. The device setup library is a separate native dependency - https://github.com/particle-iot/spark-setup-android, more info here https://docs.particle.io/reference/android/#android-device-setup-library. I don't have much experience with android so wasn't comfortable attempting to add that in. I do require it for the project I'm working on though. Is this something that you will be interested in adding?

File: src/particle.android.ts (hunk 1 above)
Comment (markoImake) [reply src/particle.android.ts]: Yes we do need that for the other login method too. Forgot to add it sorry. We could start the eventworker lazily. It needs a list of cached devices identical to the main worker though (I couldn't figure out a way to 'share' the cached devices between the workers), so maybe it should be started when calling listDevices for the first time, just before calling eventWorker.postMessage({action: "listDevices"});
What's your thoughts?

File: src/particle.android.ts (hunk 2 above)
Comment (EddyVerbruggen) [reply src/particle.android.ts:169]: Ah, understood, fair enough! I always strive for x-platform feature parity so I might give it a stab! Not this week though, I'm swamped.

Comment (EddyVerbruggen) [reply src/particle.android.ts:169]: .. that being said, that's not blocking for merging this PR of course.

File: src/particle.android.ts (hunk 1 above)
Comment (EddyVerbruggen) [reply src/particle.android.ts]: Yep, agreed! And indeed, sharing native stuff between workers is not currently possible in NativeScript.

Comment (markoImake) [reply src/particle.android.ts]: I have updated this as discussed.

File: src/particle.android.ts (hunk 2 above)
Comment (markoImake) [reply src/particle.android.ts:169]: Ok sounds good. Do I need to do anything further to get the PR merged?

Comment (EddyVerbruggen) [reply src/particle.android.ts:169]: Nope, this is perfect, I'll try to merge it asap.

**Pull request merged**